Configure Scrape Interval and Timeout Via Relabeling

LeviHarrison opened the pull request:
<!--
    Don't forget!
    
    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.
    
    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.
    
    - No tests are needed for internal implementation changes.
    
    - Performance improvements would need a benchmark test to prove it.
    
    - All exposed objects should have a comment.
    
    - All comments should start with a capital letter and end with a full stop.
 -->

This PR adds the labels `__scrape_interval__` and `__scrape_timeout__` to configure the scrape interval and timeout of a target.

Fixes #8895

LeviHarrison commented:
Just realized I forgot the UI component of this, I'll get to it later.

LeviHarrison commented:
So the current scrape interval and timeout are already shown in the UI through their labels when hovering over the `Labels` box. It might be a good idea to add a separate column (or columns) for them though because you have to know where to look. Thoughts?

roidelapluie commented:
My idea was to have them with a tooltip on the last scrape duration

LeviHarrison commented:
Ok, I added that in.

juliusv commented on scrape/scrape.go:
```diff
@@ -1452,6 +1473,24 @@ loop:
 		}
 
+		// Update the scrape loop's interval and timeout from their respective labels.
+		if intervalLabelValue := lset.Get(model.ScrapeIntervalLabel); intervalLabelValue != "" {
```
I'm just taking a very first glance here and have to run for now, but I noticed that something's being conflated here that probably shouldn't: this is looking at scraped metric labels, and not at target labels, right? As far as I understand, the issue was about being able to control these durations via target relabeling, not via metrics relabeling?

juliusv replied:
Now reading that issue it doesn't make that 100% clear, but I'm pretty sure that's what it meant.

LeviHarrison replied:
Yes, the issue was to control via target relabeling, and I believe this still does. The `lset` here is generated in the `populateLabels` function, where other target labels such as `job` and `__metrics_path__` are set.

https://github.com/prometheus/prometheus/blob/693c948c593a40c5ede20c3ab9dc115717baaac3/scrape/target.go#L325-L430

They are mixed with metric labels, which is why `__scrape_interval__` and `__scrape_timeout__` are accessed like that. They're still target labels though, and if you look in the UI they show up in the target labels popup.

juliusv replied:
:exploding_head: I guess you're right. Ok, I'll leave this to someone more into the current scrape code, sorry :laughing:

Though somehow I assumed that a new target would be created anyway in `scrapePool.sync()` because the hash would be different, but probably I'm missing more stuff without looking really closely.

LeviHarrison replied:
No worries it took me a little while to sort all this out too. I still don't totally know what's going on 🤣

roidelapluie replied:
This is indeed not the correct place. The loop should have the correct interval and timeout directly, in reload and in create.

LeviHarrison replied:
Yup, I see my fault now. Sorry, Julius!

roidelapluie reviewed (commented):
Julius is right, the loop should be created with the correct interval.

Also, instead of changing all the tests, this is worth it a dedicated test on its own. Current tests should not be affected.

Thanks!

LeviHarrison commented:
I still need to add stuff in reload tests, but here's my plan now. In `sync`, the interval and timeout of the scrape loop are set directly from the `__scrape_interval__` and `__scrape_timeout__`, which themselves are initialized with the config values and remain that value unless changed by relabeling. In `reload`, we first check if the interval and timeout differ from the previous config. If they have, that means they've been set with relabeling and we should keep them that way, even if the new config specifies different values. If they haven't, we just use the new config.

roidelapluie commented:
Did you notice `func populateLabels(lset labels.Labels, cfg *config.ScrapeConfig) (res, orig labels.Labels, err error) {` ? That where we should set the labels.

LeviHarrison commented:
Yup, that's where I'm setting them.

https://github.com/prometheus/prometheus/blob/c44283b420173579e8781c3314247c6434e34048/scrape/target.go#L328-L336

roidelapluie commented on scrape/scrape.go:
```diff
@@ -509,6 +514,21 @@ func (sp *scrapePool) sync(targets []*Target) {
+			intervalLabel := t.labels.Get(model.ScrapeIntervalLabel)
+			interval, err := time.ParseDuration(intervalLabel)
+			if err != nil {
+				level.Error(sp.logger).Log("msg", "Error parsing interval label", "err", err, "value", intervalLabel)
```
should't we call the setForcedError instead, so the target is not scraped?

LeviHarrison replied:
Yeah, probably.

LeviHarrison replied:
Actually, `setForcedError` is called on a scrape loop, which doesn't exist yet and won't exist because of the `continue`. Also, there won't be any errors here anyway, the labels have already been parsed and checked, the only reason to parse them again here is to use the durations.

LeviHarrison replied:
I guess we could save the durations we parse in `populateLabels`, because we're just repeating the same here, but that would require additional fields (which actually might have to exist anyway for the UI).

roidelapluie replied:
The goal is to go with the default scrape duration and start the for loop with the forced error.

LeviHarrison replied:
Ok, so if an error is returned from parsing (which it should never because it's already been parsed), we should fall back on the default scrape interval and duration, and set a forced error. Correct?

roidelapluie replied:
Yes, so it is like if you set a wrong __metrics_path__ or __address__, creates up = 0

LeviHarrison commented:
On reload, there can be a new scrape interval and timeout, so we want to apply that to the targets, unless the targets have set their own interval and timeout through relabeling. Then, we should just leave them alone. My problem is we have no current way of knowing if the targets have set their scrape interval and timeout through relabeling. We could just check if they still have the same timeout and interval as the previous configuration, but then what if they had set their interval and timeout to the same values of that configuration. We would then want to keep those values because they're what the target requested, but label-wise, it is undistinguished from other targets which are using the timeout and interval from the configuration. I think the only option would be to run `populateLabels` (which performs relabeling) again during a reload, but that seems wasteful.

roidelapluie commented:
> On reload, there can be a new scrape interval and timeout, so we want to apply that to the targets, unless a target has set its own interval and timeout through relabeling. Then, we should just leave it alone. My problem is we have no current way of knowing if the targets have set their scrape interval and timeout through relabeling. We could just check if they still have the same timeout and interval as the previous configuration, but then what if they had set their interval and timeout to the same values of that configuration. We would then want to keep those values because they're what the target requested, but label-wise, it is undistinguished from other targets which are using the timeout and interval from the configuration. I think the only option would be to run `populateLabels` (which performs relabeling) again during a reload, but that seems wasteful.

I propose we work on this specific problem separately. Maybe open an issue so we do not forget.

LeviHarrison commented:
I think this is critical to the feature though. Without a solution, either we completely forget the relabeled interval and timeout on reload, or we don't apply the configurations new interval and timeout to any target on reload (this is probably worse). Either way, it's a bug.

juliusv commented:
Just a drive-by question:

> unless a target has set its own interval and timeout through relabeling.

Does this mean this PR does implement custom scrape intervals coming from actual instrumentation labels now? Or with "set its own interval", do you just mean that an individual target's target labels have an interval setting?

LeviHarrison commented:
This PR aims to allow one to set the scrape interval and target via relabeling targets labels coming from a target.

juliusv commented:
Ok, I guess it's the "coming from a target" that is ambiguous: I guess you mean coming from the target object, not from the actual scraped metrics, right?

LeviHarrison commented:
Yup, sorry if that was unclear

juliusv commented:
Yup, ok :) Just wanted to make sure!

roidelapluie commented on scrape/scrape.go:
```diff
@@ -509,6 +514,21 @@ func (sp *scrapePool) sync(targets []*Target) {
+			timeout, err := time.ParseDuration(timeoutLabel)
+			if err != nil {
+				level.Error(sp.logger).Log("msg", "Error parsing timeout label", "err", err, "value", timeoutLabel)
+			}
```
We should check that the timeout <= scrape interval; here is e.g. what we have now:

parsing YAML file prometheus.yml: global scrape timeout greater than scrape interval

LeviHarrison replied:
I'll add that in the `populateLabels` function in `target.go`, which is where all the label checking is done. They're only parsed here again because there's nowhere to store the parsed value in the intermediate.

LeviHarrison replied:
Theoretically, there should never be an error here.

roidelapluie commented:
It seems like this requires documentation, plus the check for the interval smaller than the timeout.

Additionally, let's mark this as experimental, in case we change some semantics later.

LeviHarrison commented:
Yeah, I started some changes but have been neglecting this for the past few days. We still need to figure out the reloading issue too.

LeviHarrison commented:
Actually, I think it might be necessary to run relabeling on config reload. Because users could theoretically set different actions for different values of `__scrape_interval__` and `__scrape_timeout__`, we'd have to rerun their relabel configs when we change them.

roidelapluie commented:
If the base values change, then reload is not called for that target, a new one is created, if I am correct.

roidelapluie commented:
Oh, I see what you mean. Yes, there is something to investigate here. While investigating, we should cover blind spots with additional tests.

LeviHarrison commented:
Yeah, don't think a new target is created when the interval or timeout is changed.

https://github.com/prometheus/prometheus/blob/fa6b2897f0db5844cb7a2c6b4f17f3bd5a3deed4/scrape/scrape.go#L410-L439

LeviHarrison commented:
I've added addressed some of @roidelapluie's review comments and added relabeling on reload, as well as tests for it. I'm going to ignore the UI component and the documentation for now until we can agree on this part of the implementation.

LeviHarrison commented on scrape/target.go:
```diff
@@ -346,8 +355,12 @@ func populateLabels(lset labels.Labels, cfg *config.ScrapeConfig) (res, orig lab
 
-	preRelabelLabels := lb.Labels()
-	lset = relabel.Process(preRelabelLabels, cfg.RelabelConfigs...)
+	return processLabels(lb.Labels(), cfg)
+}
+
+// processLabels runs relabeling and checks the output.
+func processLabels(preRelabelLabels labels.Labels, cfg *config.ScrapeConfig) (res, orig labels.Labels, err error) {
```
I split off the relabeling part of `populateLabels` into a new function because on reload we only need that part.

LeviHarrison commented on scrape/target.go:
```diff
@@ -273,6 +283,31 @@ func (t *Target) Health() TargetHealth {
 
+// IntervalAndTimeout returns the interval and timeout derived from
+// the targets labels.
+func (t *Target) IntervalAndTimeout(prevInterval, prevDuration time.Duration) (time.Duration, time.Duration, error) {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+
+	intervalLabel := t.labels.Get(model.ScrapeIntervalLabel)
+	interval, err := model.ParseDuration(intervalLabel)
+	if err != nil {
+		return prevInterval, prevDuration, errors.Errorf("Error parsing interval label %q: %v", intervalLabel, err)
+	}
+	timeoutLabel := t.labels.Get(model.ScrapeTimeoutLabel)
+	timeout, err := model.ParseDuration(timeoutLabel)
+	if err != nil {
+		return prevInterval, prevDuration, errors.Errorf("Error parsing timeout label %q: %v", timeoutLabel, err)
+	}
+
+	return time.Duration(interval), time.Duration(timeout), nil
+}
```
I still don't exactly like how this function takes the previous interval and duration as arguments. The idea was to skip an error check (in case of a cosmic ray) by returning the old values on error, and it is handy:

https://github.com/prometheus/prometheus/blob/315031aa6f376606567375d511bf67982a9535ac/scrape/scrape.go#L523-L536

but at the cost of things being weird if this function is used in any other scenario (which it isn't currently).

roidelapluie replied:
let's unexport this function

roidelapluie replied:
Instead of pref, should we call this default?

roidelapluie replied:
friendly ping :)

LeviHarrison replied:
Sorry, I was waiting on a response to https://github.com/prometheus/prometheus/pull/8911#discussion_r673055312

LeviHarrison commented:
We might want to consider moving the interval and timeout somewhere out of the tooltip on the last scrape duration because you'd never know it's there. Even I forgot for a second where it was just now. Maybe it gets its own box or is nestled into the same one as the last scrape duration.

LeviHarrison commented:
> It seems like this requires documentation

Where would we put it?:

Related: https://github.com/prometheus/docs/issues/1951#issuecomment-875192148

roidelapluie commented on scrape/scrape.go:
```diff
@@ -510,6 +517,12 @@ func (sp *scrapePool) sync(targets []*Target) {
+			// so whether changed via relabeling or not, they'll exist and hold the correct values
+			// for every target.
+			var err error
+			interval, timeout, err = t.IntervalAndTimeout(interval, timeout)
```
If we set it and use it, we should not error.

roidelapluie commented on scrape/target.go:
```diff
@@ -390,6 +427,34 @@ func populateLabels(lset labels.Labels, cfg *config.ScrapeConfig) (res, orig lab
 		return nil, nil, err
 	}
 
+	var interval string
```
Can we do something to avois this code duplication, or does it end being more complex?

LeviHarrison replied:
Over here we do more than just parsing, we also check to make sure the value is valid, so probably not.

roidelapluie commented on web/api/v1/api.go:
```diff
@@ -768,6 +768,9 @@ type Target struct {
 	LastScrapeDuration float64             `json:"lastScrapeDuration"`
 	Health             scrape.TargetHealth `json:"health"`
+
+	Interval string `json:"interval"`
```
Can we call that scrape_interval and scrape_timeout, like the config options?

LeviHarrison replied:
I was thinking that after validating in `populateLabels`, we could then modify the actual config fields, which would allow us to access the values here and `sync` without any additional parsing. WDYT?

roidelapluie replied:
yes that looks like a simple approach, it makes populateLabels a bad name then, but I do not have a better name. let's make sure that this is documented in populateLabels help.

LeviHarrison replied:
After trying to actually implement this, I've realized that it's not viable because the config passed to `populateLabels` is for the entire target group, not just the individual targets which we want to modify. This leaves us with two options:

1. Keep the existing system where we store the interval and timeout in labels and re-parse to access using `IntervalAndTimeout`. This has the benefit of complying with how we store other values like `job`, `address`, and `scheme`.
2. Store the values in standalone fields in the `Target`. This way, we don't have to parse them again when accessing.

roidelapluie replied:
Let's go for 1.

roidelapluie reviewed (commented):
Thanks, I have let a few comments. Can you please also adapt the documentation?

LeviHarrison commented:
> Can you please also adapt the documentation?

> > It seems like this requires documentation
> 
> 
> 
> I agree, but where would we put it?
> 
> 
> 
> Related: https://github.com/prometheus/docs/issues/1951#issuecomment-875192148

roidelapluie commented:
Docs should go here https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config

LeviHarrison commented:
@roidelapluie friendly ping :)

roidelapluie commented on docs/querying/api.md:
```diff
@@ -511,6 +513,8 @@ $ curl http://localhost:9090/api/v1/targets
           "__address__": "127.0.0.1:9100",
           "__metrics_path__": "/metrics",
           "__scheme__": "http",
+          "__scrape_interval": "1m",
```
```suggestion
          "__scrape_interval__": "1m",
```

roidelapluie commented on docs/querying/api.md:
```diff
@@ -511,6 +513,8 @@ $ curl http://localhost:9090/api/v1/targets
           "__metrics_path__": "/metrics",
           "__scheme__": "http",
+          "__scrape_interval": "1m",
+          "__scrape_timeout": "10s",
```
```suggestion
          "__scrape_timeout__": "10s",
```

roidelapluie commented on scrape/scrape_test.go:
```diff
@@ -484,8 +488,8 @@ func TestScrapePoolAppender(t *testing.T) {
-	interval, _ := model.ParseDuration("500ms")
-	timeout, _ := model.ParseDuration("1s")
+	interval, _ := model.ParseDuration("1s")
+	timeout, _ := model.ParseDuration("500ms")
```
interval < timeout should be rejected

LeviHarrison replied:
But `1s` > `500ms`???

https://github.com/prometheus/prometheus/blob/ca526806485de720daa513b472bed68ea6c96033/scrape/target.go#L454-L456

roidelapluie replied:
oops you are correct

mea culpa

roidelapluie commented on docs/configuration/configuration.md:
```diff
@@ -2172,6 +2172,9 @@ it was not set during relabeling. The `__scheme__` and `__metrics_path__` labels
 are set to the scheme and metrics path of the target respectively. The `__param_<name>`
 label is set to the value of the first passed URL parameter called `<name>`.
 
+The `__scrape_interval__` and `__scrape_timeout__` labels are set the target's
```
```suggestion
The `__scrape_interval__` and `__scrape_timeout__` labels set the target's
```

LeviHarrison replied:
I believe the proper use is "The `__scrape_interval__` and `__scrape_timeout__` labels are set  _to_ the target's..."

roidelapluie replied:
yeah

roidelapluie commented on scrape/manager_test.go:
```diff
@@ -201,14 +235,102 @@ func TestPopulateLabels(t *testing.T) {
+			resOrig: nil,
+			err:     "scrape timeout cannot be 0",
+		},
+		// Timeout less or equal to interval.
```
```suggestion
		// Timeout less than interval.
```

We allow equal AFAIK

roidelapluie reviewed (commented):
very last comments

roidelapluie commented:
:100: thanks!!!!

LeviHarrison commented:
Thanks so much for all your time reviewing! This one took a while.

roidelapluie commented:
@beorn7 FYI this feature is marked experimental and does not have a feature flag.

The reason for marking as experimental is the refusal to scrape when the scrape timeout > scrape interval. While I think we do the right thing here, I do not exclude that we revisit this, change or improve in the future.

beorn7 commented:
In general, I would prefer to use feature flags for all experimental features (where that's reasonably possible), for simple consistency. If we only expect minor changes to the semantics, I can also live without the feature flag as we don't expect to change the behavior fundamentally or completely rip out the feature, i.e. most users of the experimental feature will probably be able to continue to use it in the way they use it for the lifetime of v2.x.

But as said, I would have preferred a feature flag here (but changing it now wouldn't be worth the hassle).

LeviHarrison commented:
> The reason for marking as experimental is the refusal to scrape when the scrape timeout > scrape interval.

Perhaps we should indicate that in the docs then.
